### PR TITLE
Fixes: Negative indexing for Array is inconsistent #14678

### PR DESCRIPTION
**Pull request by yeshsurya**
Fixes "Negative indexing for Array is inconsistent #14678"
When the indexes are negative initial value of real_index is not zero, it should increase. Scenario where index is positive will work as it was previous to fix.

### CODE REVIEW

**Comment by Abdullahjavednesar** [comment]
@yeshsurya please add a test case.

File: sympy/tensor/array/ndim_array.py [hunk 1]
```diff
@@ -79,6 +79,8 @@ def _parse_index(self, index):
         for i in range(self._rank):
             if index[i] >= self.shape[i]:
                 raise ValueError('Index ' + str(index) + ' out of border')
+            if(index[i]<0):
+                real_index+=1
```
Comment (sidhantnagpal) [inline sympy/tensor/array/ndim_array.py]: Spaces should be added before and after `<`, similarly for `+=`.

**Comment by yeshsurya** [comment]
@Abdullahjavednesar , Could you please suggest the where to add the test case. Is it like new test case file has be added or we need to add in any existing test case file. Also if a new test case file has to be created then is it in "tensor/tests" or "tensory/array/tests"

**Comment by sidhantnagpal** [comment]
A new file might be required `tensor/array/tests/test_ndim_array.py` (as it currently does not exist).
See [test](https://github.com/sympy/sympy/blob/master/sympy/tensor/array/tests/test_ndim_array_conversions.py) file for the format.

**Comment by Abdullahjavednesar** [comment]
May be @Upabjojr can help here.

**Comment by Upabjojr** [comment]
There is no boundary check for the negative indices. Do we need one?

**Comment by yeshsurya** [comment]
@Upabjojr ,You are talking about whether to fix the issue or if test case is required ?

**Comment by Upabjojr** [comment]
Test is required. Issue may be fixed here or in another PR. But better here.

**Comment by yeshsurya** [comment]
Added tests related to Issue #14679

**Comment by sidhantnagpal** [comment]
> Fixes "Negative indexing for Array is inconsistent #14678"

To use the autoclose format for the issue use keywords immediately before the issue number: "Fixes #14678" or "Closes #14678". 
See [here](https://help.github.com/articles/closing-issues-using-keywords/) for more information.

File: sympy/tensor/array/tests/test_ndim_array.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+from sympy import Array
+
+
+def test_array_negative_indices():
+ test_array = Array([[1,2,3,4,5],[6,7,8,9,10]])
```
Comment (Upabjojr) [inline sympy/tensor/array/tests/test_ndim_array.py]: We use the standardized 4 spaces indentation in SymPy.

By the way, maybe this should be put into the other test files? I don't see the point of creating a new file.

Comment (yeshsurya) [reply sympy/tensor/array/tests/test_ndim_array.py]: @sidhantnagpal felt tests were to be in new file.So, new file was added. @Upabjojr , could you please help me in the build failure, for there is failure in first assertion. test_array[:,-1] is [10,5]. However the assertion for the first one is failing, not able to figure out the reason.

Comment (Upabjojr) [reply sympy/tensor/array/tests/test_ndim_array.py]: did you debug?

Comment (yeshsurya) [reply sympy/tensor/array/tests/test_ndim_array.py]: I tested in my local machine, The assertion is "True" for the line to which build is marking as failure.

**Comment by czgdp1807** [comment]
@yeshsurya Are you still interested to work on it?

**Comment by Upabjojr** [comment]
This definitely needs to be merged somehow.

File: sympy/tensor/array/ndim_array.py [hunk 3]
```diff
@@ -79,6 +79,8 @@ def _parse_index(self, index):
         for i in range(self._rank):
             if index[i] >= self.shape[i]:
                 raise ValueError('Index ' + str(index) + ' out of border')
+            if index[i] < 0:
+                real_index += 1
```
Comment (Upabjojr) [inline sympy/tensor/array/ndim_array.py]: You need to check if the index is too small (below the dimension limit).

**Comment by Upabjojr** [comment]
Continue in https://github.com/sympy/sympy/pull/17669

**Pull request closed without merging**